L73-java-binderchannel/wireformat.md

```diff
@@ -0,0 +1,200 @@
+# The wire format of BinderTransport.
+
+All communication occurs via one-way binder
+[transactions](https://developer.android.com/reference/android/os/Binder#transact\(int,%20android.os.Parcel,%20android.os.Parcel,%20int\))
+which are made up of an integer “code”, and a [Parcel] containing arbitrary
+data.
+
+Three types of binder are involved in our implementation.
+
+*   The _endpoint binder_, owned by the BinderServer or OnDeviceServerEndpoint
+    instance, and returned from the host Android service's onBind method.
+*   The _client binder_, owned by the ClientTransport implementation within the
+    Channel.
+*   The _server binder_, owned by the ServerTransport implementation.
+
+A single endpoint binder exists for each BinderServer or OnDeviceServerEndpoint.
+A client binder / server binder pair is created for each transport pair.
+
+## Transaction Types
+
+There are two types of transactions, control and stream.
+
+### Transport-level Control Transactions
+
+For general transport lifecycle, using one of 1000 reserved transaction codes.
+
+*   Setup Transport
+    *   Transaction code: BinderTransport.SETUP_TRANSPORT (1)
+    *   From ClientTransport to host binder, and from ServerTransport to client
+        binder.
+*   Shutdown Transport
+    *   Transaction code: BinderTransport.SHUTDOWN_TRANSPORT (2)
+    *   From ClientTransport to server binder, or ServerTransport to client
+        binder.
+*   Acknowledge bytes.
+    *   Transaction code: BinderTransport.ACKNOWLEDGE_BYTES (3)
+    *   From either transport to its peer's binder. Reports reception of
+        transaction data for flow control.
+*   Ping
+    *   Transaction code: BinderTransport.PING (4)
+    *   Currently only sent from ClientTransport to server binder.
+*   Ping Response:
+    *   Transaction code: BinderTransport.PING_RESPONSE (5)
+    *   From SeverTransport to client binder or ClientTransport to server binder,
+        in response to a ping message.
+
+If either side of a transport receives an unrecognized control transacton,
+it will shutdown the transport gracefully.
+
+#### Transaction Data
+
+```
+major version = int;
+protocol extension flags = int;
+num bytes = int;
+ping id = int;
+
+setup transport transaction = major version, binder, [protocol extension flags];
+shutdown transport transaction = nothing;
+acknowledge bytes transaction = num bytes;
+ping transaction = ping id;
+ping response transaction = ping id;
+```
+
+During transport setup, if a server receves a setup transaction with a higher
+protocol version, it should respond with the protocol version _it_ supports.
+
+Both client and server transport may also include protocol extension flags at
+the end of their setup transport transaction. This is reserved for potential
+future protocol extensions, though no flags are currently specified. Since some
```

Comment:
I'm not sure the "Since some features may..." sentence actually adds anything, since the expected behavior is _always_ specific to that one flag. So you may just drop it.

It is probably more important to say that when unknown flags are received, they should be ignored.

Reply:
Yeah, that sentence annoyed me while writing it. Happy to drop, and added the 'ignored' part.